### ISSUE TITLE
Debug assertion failure: AlongStep.hh:129: `mfp > 0`

### ISSUE THREAD

**Issue by amandalund** [celeritas-project/celeritas#615]
#### Failure:
```json
    "failure": {
     "condition": "mfp > 0",
     "file": "/home/alund/celeritas_project/celeritas/src/celeritas/global/alongstep/AlongStep.hh",
     "line": 129,
     "type": "DebugError",
     "which": "internal assertion failed"
    }
```
#### Regression problem:
cms2018+field+msc-vecgeom-cpu

#### Diagnosis:
The positron step is limited by the distance to boundary = 8.64658290263130985e-02 (with distance to interaction = 9.46711477923071754e-02). Since it's a geo-limited step, `UrbanMscScatter` recalculates the true path length, with length = 9.48569009850495776e-02, which is then clamped to the interaction length. Because the step length is equal to the interaction length, the mfp is reduced to zero, but the action is still boundary crossing.

**Comment by sethrj** [comment]
Additional failures from wildstyle in `cms2018+field+msc-vecgeom-cpu`:
```json
"failure": {
 "condition": "mfp > 0",
 "context": {
  "dir": [
   0.34075079638596456,
   0.11890851532618418,
   0.9326036991912768
  ],
  "energy": [
   9.331117982852458,
   "MeV"
  ],
  "event": 4,
  "label": "along-step-uniform-msc",
  "num_steps": 3,
  "parent": 1681145,
  "particle": 2,
  "pos": [
   -100.696255136249,
   87.18868463055634,
   80.93651914108645
  ],
  "thread": 593,
  "track": 1681186,
  "type": "KernelContextException",
  "volume": 2705
 },
 "file": "/home/s3j/.local/src/celeritas/src/celeritas/global/alongstep/AlongStep.hh",
 "line": 129,
 "type": "DebugError",
 "which": "internal assertion failed"
}
```
and
```json
"failure": {
 "condition": "mfp > 0",
 "context": {
  "dir": [
   0.24278444495415058,
   0.4229572039704155,
   0.8730194252694686
  ],
  "energy": [
   9.916999303797795,
   "MeV"
  ],
  "event": 3,
  "label": "along-step-uniform-msc",
  "num_steps": 3,
  "parent": 28006,
  "particle": 2,
  "pos": [
   65.39978882143026,
   55.44444787294294,
   320.33029114931753
  ],
  "thread": 3255,
  "track": 30777,
  "type": "KernelContextException",
  "volume": 3120
 },
 "file": "/home/s3j/.local/src/celeritas/src/celeritas/global/alongstep/AlongStep.hh",
 "line": 129,
 "type": "DebugError",
 "which": "internal assertion failed"
}
```

**Comment by amandalund** [comment]
Digging into this failure more (on `testem3-flat+msc-orange-cpu`), I found the following:
 - The assertion is triggered more often when the number of bins per decade is increased (zero times in this test with Geant4's default of 7 bins per decade, but 90 times with the regression problem's 56 bins per decade)
 - In those 90 cases, the particle was always a positron, was always in the gap (Pb), and alway had a pre-step energy a little over 10 MeV

Looking into how the Urban MSC cross sections are calculated in Geant4 (`G4UrbanMscModel::ComputeCrossSectionPerAtom`), it appears that:
 - Two different methods are used to calculate the cross section, determined by an energy threshold of 10 MeV
 - Below 10 MeV, correction factors tabulated in Z and energy for both electrons and positrons are applied to a simple relativistic screened Rutherford cross section
 - Above 10 MeV, the cross section is calculated using the first transport cross sections for electrons and a high energy correction factor tabulated in Z

The following plots show the e+/e- Urban MSC cross sections on both the coarse and fine energy grids for the two materials in the problem:
![msc-xs-ar](https://user-images.githubusercontent.com/6426426/217689317-92a3688e-a1e7-4073-9ea0-603cb2799eac.png)
![msc-xs-pb](https://user-images.githubusercontent.com/6426426/217689324-7536f7bd-8f1c-417c-b8a6-82108c21e536.png)

Because the cross section above the 10 MeV threshold is the same for electrons and positrons, there is a large, unphysical discontinuity in the positron cross section.

If a positron's energy is just above the 10 MeV threshold, a small change in energy over the step will result in an erroneously large change in the transport mfp (which will be amplified on a finer energy grid). In the true path --> geom path transformation in the Urban MSC step limit calculation, this can lead to a geometrical path length that's too large. The final geom path --> true path conversion can then produce a true path larger than the interaction length.

**Comment by sethrj** [comment]
We discussed the MFP issue at length today. I believe @whokion is right in that since we've done detailed comparisons with Geant4 for our MSC calculations, we're not doing something "wrong" (or at least something "different" or "inconsistent"). However, our persnickety assertions mean we could see problems that Geant4 glosses over: for example, we know that with energy loss fluctuations enabled, both Celeritas and Geant4 can artificially bias particles to lose all their energy on material boundaries (leading to annihilation gammas tending to come from the hard edge of a volume rather than in the bulk volume near the boundary).

That being said: it's clear from the pathology that the erroneously large geometry path estimation results from the branch https://github.com/celeritas-project/celeritas/blob/1954a8a29fb59798f910ea3f9b45f815399af905/src/celeritas/em/msc/UrbanMscStepLimit.hh#L301 . This calculates geometrical path length $l$ from the physical path length $s$ using:

$$ \alpha = \frac{\lambda_i - \lambda_f}{s \lambda_f} $$

where $\lambda_i = \lambda(E(0))$ and $\lambda_f= \lambda(E(s))$ are the transport (MSC) mean free paths at the initial and final energies along the step $s$. The corresponding geometric path length is derived in section 8.1.3 of the Geant4 physics manual, and (8.10) is the final value of `result.geom_path` in that branch. Defining $w = 1 + 1/(\alpha \lambda_i)$, this equation simplifies to:

$$ l = \frac{1 - (\lambda_f / \lambda_i)^w }{\alpha w} .$$

_However_, this equation is based on the assumption that the cross section varies _linearly_ over the step:

$$ \lambda(E(x)) = \lambda(E_0) (1 - \alpha x) .$$

Although the physics manual asserts that this is "not a crude approximation," it is unfortunately a very crude approximation when the actual cross section has sudden discontinuities in it. However, if the energy loss over a step is small and the cross sections are smooth, it's a great approximation.

**Comment by sethrj** [comment]
OK looking into this a *little* more, I think there may be an asymmetry between `UrbanMscStepLimit::calc_geom_path` and `UrbanMscScatter::calc_true_path`. I don't think the last conditional (the linear lambda approximation) is exactly inverted: maybe we need to use the true range calculation instead of truncating to `range`?

**Comment by amandalund** [comment]
> I don't think the last conditional (the linear lambda approximation) is exactly inverted: maybe we need to use the true range calculation instead of truncating to `range`?

@sethrj this may be relevant for #620 

Actually, even though the true path --> geom path conversion is done using PRM eq. 8.10, since `alpha` will be negative (`lambda1` will be much larger than `lambda`) I think we always end up doing the geom path --> true path conversion in `UrbanMscScatter::calc_true_path` using the equation for the case where the step is small.

**Comment by sethrj** [comment]
Oh yeah, I got these two issues confused.

WRT alpha: I think the `alpha < 0` case is *exclusively* to check for the `alpha = -1` flag for the super-short MFP cases? The initial mean free path is longer than the final as long as the cross section decreases with energy (decreasing XS with energy → increasing mfp with energy → decreasing mfp over step length → $\lambda_i > \lambda_f$ → $\alpha > 0$).

**Comment by amandalund** [comment]
I think that's backwards? If the cross section decreases as the energy decreases, then the mfp increases as energy decreases, and the final mfp is larger than the initial mfp and alpha is negative.

**Comment by sethrj** [comment]
Uh. OK I was thinking higher-energy would have longer MFP than lower energy because generally this is the case... Then I looked at your cross section plots which I now realize are *increasing* with energy, and I got confused by the back-and-forth of increasing energy vs increasing travel along the step.

Do your cross section plots above account for the extra $(E / \mathrm{MeV})^2$ that's multiplied into the MSC params?

**Comment by amandalund** [comment]
No, the extra E^2 is still in there (so that label is wrong).

**Comment by amandalund** [comment]
Yep, with the correct units:
![msc-xs-pb2](https://user-images.githubusercontent.com/6426426/218223586-670fbef9-1780-43dd-b0ab-51350f8e8f9e.png)

**Comment by amandalund** [comment]
And if the `alpha < 0` case really is just to check for the `-1` flag for super-short mfp, then maybe the fix is just to change that to explicitly check `alpha == -1` and let these negative alpha cases caused by the xs discontinuity use the correct g --> t transformation?

**Comment by sethrj** [comment]
Wow, so alpha < 0 might really be the problem for that (or perhaps a second problem *besides* the linearity). I'm actually half a step ahead of you, because as I was trying to unravel this and deduced that `alpha < 0` really meant `is_flag(alpha)`, I put in that fix (1b3427992debc18ad26061dfbe24ffb10f85bed2) in an hour before you commented 😉 ... I just didn't finish the pull request last night.

**Comment by sethrj** [comment]
By the way, the `clamp` in `UrbanMscScatter::calc_true_path` isn't tested in UrbanMsc: as it happens, the entire `calc_true_path` functionality is untested...
